src/util.ts

```diff
@@ -41,6 +42,38 @@ interface PackageJson {
   version: string;
 }
 
+export interface ClassOf<T, Config> {
+  new(logger: Logger, config: Config): T;
+  prototype: T;
+  name: string;
+}
+
+/**
+ * An class that provides access to a singleton.
+ * Instances of this type should only be constructed in module scope.
+ */
+export class Singleton<T, Config> {
+  private singleton: T|null = null;
+
+  constructor(private implementation: ClassOf<T, Config>) {}
```

Comment:
`private readonly implementation`?

Reply:
In a test we specifically want to override the `implementation` field with a dummy class constructor. Being able to write (for example) `traceWriterSingleton['implementation'] = TestTraceWriter` is the easiest way to do this, in my opinion.

Reply:
I see. How is it possible to overwrite the `implementation` field when it is a private field? Doesn't the compiler prevent that?